File: tools/generate-sitemap.php
#!/usr/bin/env php
<?php
/**
 * Sitemap Generator for 3D Print Pro
 * 
 * Генерирует sitemap.xml для статического сайта с учетом всех публичных страниц и якорей
 * Запуск: php tools/generate-sitemap.php [base_url]
 * 
 * @version 1.0.0
 */

// Определяем базовый URL
$baseUrl = $argv[1] ?? 'https://3dprint-omsk.ru';
$baseUrl = rtrim($baseUrl, '/');

// Определяем структуру сайта
$urls = [
    // Главная страница (с максимальным приоритетом)
    [
        'loc' => $baseUrl . '/',
        'changefreq' => 'daily',
        'priority' => '1.0',
        'lastmod' => date('Y-m-d')
    ],
    
    // Основные разделы через якоря
    [
        'loc' => $baseUrl . '/#home',
        'changefreq' => 'daily',
        'priority' => '1.0',
        'lastmod' => date('Y-m-d')
    ],
    [
        'loc' => $baseUrl . '/#services',
        'changefreq' => 'weekly',
        'priority' => '0.9',
        'lastmod' => date('Y-m-d')
    ],
    [
        'loc' => $baseUrl . '/#calculator',
        'changefreq' => 'weekly',
        'priority' => '0.8',
        'lastmod' => date('Y-m-d')
    ],
    [
        'loc' => $baseUrl . '/#portfolio',
        'changefreq' => 'weekly',
        'priority' => '0.9',
        'lastmod' => date('Y-m-d')
    ],
    [
        'loc' => $baseUrl . '/#about',
        'changefreq' => 'monthly',
        'priority' => '0.7',
        'lastmod' => date('Y-m-d')
    ],
    [
        'loc' => $baseUrl . '/#contact',
        'changefreq' => 'monthly',
        'priority' => '0.8',
        'lastmod' => date('Y-m-d')
    ],
];

// Проверка наличия файла index.html для определения даты последнего изменения
$indexFile = __DIR__ . '/../index.html';
if (file_exists($indexFile)) {
    $lastModified = date('Y-m-d', filemtime($indexFile));
    foreach ($urls as &$url) {
        $url['lastmod'] = $lastModified;
    }
}

// Генерация XML
$xml = new DOMDocument('1.0', 'UTF-8');
$xml->formatOutput = true;

$urlset = $xml->createElement('urlset');
$urlset->setAttribute('xmlns', 'http://www.sitemaps.org/schemas/sitemap/0.9');
$xml->appendChild($urlset);

foreach ($urls as $urlData) {
    $url = $xml->createElement('url');
    
    $loc = $xml->createElement('loc', htmlspecialchars($urlData['loc']));
    $url->appendChild($loc);
    
    if (isset($urlData['lastmod'])) {
        $lastmod = $xml->createElement('lastmod', $urlData['lastmod']);
        $url->appendChild($lastmod);
    }
    
    if (isset($urlData['changefreq'])) {
        $changefreq = $xml->createElement('changefreq', $urlData['changefreq']);
        $url->appendChild($changefreq);
    }
    
    if (isset($urlData['priority'])) {
        $priority = $xml->createElement('priority', $urlData['priority']);
        $url->appendChild($priority);
    }
    
    $urlset->appendChild($url);
}

// Сохранение файла
$outputFile = __DIR__ . '/../sitemap.xml';
$xml->save($outputFile);

// Вывод информации
echo "✅ Sitemap generated successfully!\n";
echo "─────────────────────────────────────\n";
echo "Base URL: {$baseUrl}\n";
echo "Output file: {$outputFile}\n";
echo "Total URLs: " . count($urls) . "\n";
echo "Last modified: " . ($lastModified ?? date('Y-m-d')) . "\n";
echo "\n";
echo "📋 Generated URLs:\n";
foreach ($urls as $urlData) {
    echo "  - {$urlData['loc']} (priority: {$urlData['priority']})\n";
}
echo "\n";
echo "🔍 Validation:\n";
echo "  - Online: https://www.xml-sitemaps.com/validate-xml-sitemap.html\n";
echo "  - Google: https://search.google.com/search-console\n";
echo "  - Yandex: https://webmaster.yandex.ru/\n";
echo "\n";

exit(0);

File: tools/generate-sitemap.py
#!/usr/bin/env python3
"""
Sitemap Generator for 3D Print Pro

Генерирует sitemap.xml для статического сайта с учетом всех публичных страниц и якорей
Запуск: python3 tools/generate-sitemap.py [base_url]

@version 1.0.0
"""

import os
import sys
from datetime import datetime
from xml.etree.ElementTree import Element, SubElement, tostring
from xml.dom import minidom

def generate_sitemap(base_url='https://3dprint-omsk.ru'):
    """Генерирует sitemap.xml"""
    
    base_url = base_url.rstrip('/')
    
    # Определяем структуру сайта
    urls = [
        # Главная страница (с максимальным приоритетом)
        {
            'loc': f'{base_url}/',
            'changefreq': 'daily',
            'priority': '1.0'
        },
        # Основные разделы через якоря
        {
            'loc': f'{base_url}/#home',
            'changefreq': 'daily',
            'priority': '1.0'
        },
        {
            'loc': f'{base_url}/#services',
            'changefreq': 'weekly',
            'priority': '0.9'
        },
        {
            'loc': f'{base_url}/#calculator',
            'changefreq': 'weekly',
            'priority': '0.8'
        },
        {
            'loc': f'{base_url}/#portfolio',
            'changefreq': 'weekly',
            'priority': '0.9'
        },
        {
            'loc': f'{base_url}/#about',
            'changefreq': 'monthly',
            'priority': '0.7'
        },
        {
            'loc': f'{base_url}/#contact',
            'changefreq': 'monthly',
            'priority': '0.8'
        },
    ]
    
    # Проверка наличия файла index.html для определения даты последнего изменения
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.dirname(script_dir)
    index_file = os.path.join(project_root, 'index.html')
    
    if os.path.exists(index_file):
        last_modified = datetime.fromtimestamp(os.path.getmtime(index_file)).strftime('%Y-%m-%d')
    else:
        last_modified = datetime.now().strftime('%Y-%m-%d')
    
    # Добавляем lastmod ко всем URL
    for url in urls:
        url['lastmod'] = last_modified
    
    # Генерация XML
    urlset = Element('urlset')
    urlset.set('xmlns', 'http://www.sitemaps.org/schemas/sitemap/0.9')
    
    for url_data in urls:
        url = SubElement(urlset, 'url')
        
        loc = SubElement(url, 'loc')
        loc.text = url_data['loc']
        
        if 'lastmod' in url_data:
            lastmod = SubElement(url, 'lastmod')
            lastmod.text = url_data['lastmod']
        
        if 'changefreq' in url_data:
            changefreq = SubElement(url, 'changefreq')
            changefreq.text = url_data['changefreq']
        
        if 'priority' in url_data:
            priority = SubElement(url, 'priority')
            priority.text = url_data['priority']
    
    # Форматирование XML
    xml_str = minidom.parseString(tostring(urlset)).toprettyxml(indent='  ')
    
    # Сохранение файла
    output_file = os.path.join(project_root, 'sitemap.xml')
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(xml_str)
    
    # Вывод информации
    print('✅ Sitemap generated successfully!')
    print('─────────────────────────────────────')
    print(f'Base URL: {base_url}')
    print(f'Output file: {output_file}')
    print(f'Total URLs: {len(urls)}')
    print(f'Last modified: {last_modified}')
    print()
    print('📋 Generated URLs:')
    for url_data in urls:
        print(f'  - {url_data["loc"]} (priority: {url_data["priority"]})')
    print()
    print('🔍 Validation:')
    print('  - Online: https://www.xml-sitemaps.com/validate-xml-sitemap.html')
    print('  - Google: https://search.google.com/search-console')
    print('  - Yandex: https://webmaster.yandex.ru/')
    print()

if __name__ == '__main__':
    base_url = sys.argv[1] if len(sys.argv) > 1 else 'https://3dprint-omsk.ru'
    generate_sitemap(base_url)
